src/map.cpp

```diff
@@ -6999,6 +7001,173 @@ void map::rotten_item_spawn( const item &item, const tripoint &pnt )
     }
 }
 
+void map::actualize_snow( float new_snow_lvl )
+{
+    const int zmin = 0;
+    const int zmax = zlevels ? OVERMAP_HEIGHT : 0;
+    tripoint gp;
+    int &gx = gp.x;
+    int &gy = gp.y;
+    int &gz = gp.z;
+    for( gz = zmin; gz <= zmax; gz++ ) {
+        for( gx = 0; gx < my_MAPSIZE; gx++ ) {
+            for( gy = 0; gy < my_MAPSIZE; gy++ ) {
+                actualize_snow_in_submap( tripoint( gx, gy, gz ), new_snow_lvl, false );
+            }
+        }
+    }
+}
+
+void map::actualize_snow_in_submap( const tripoint &gridp, float new_snow_lvl, bool from_scratch )
+{
+    /**
+     * Takes the snow level in mm and materializes it in a submap in the form of a `fd_snow` field
+     * with different intensities that correspond to the snow depth.
+     * Unless `from_scratch` is true, tries to update the existing snow layer according to the difference
+     * between `submap::snow_level` and `new_snow_lvl`.
+     * Behavior:
+     * * when snow is updated from 0, an even layer of intensity 1 is created
+     * * same happens every 10mm of snow (all patches cleared by player are covered with intensity 1)
+     * * after 10mm of snow field intensity is upgraded pseudo-randomly until the next intensity level is reached
+     *    (see `actualize_snow_in_submap::level_to_intensity`)
+     * * same process happens backwards (intensity is downgraded) when `new_snow_lvl` is less than `submap::snow_level`
+     * * when snow is melting, evel layer of intensity 1 is preserved until snow level reaches zero (eps)
+     *
+     * Important behavior note: `actualize_snow_in_submap` should have exact same behavior regardless of whether
+     *  new_snow_level is incremented/decremented once by large amount or multiple times in tiny increments/decrements.
+     * This is important, because newly loaded and actualized submaps should look exactly like submaps that were
+     *  actively updated in the reality bubble.
+     */
+
+    if( gridp.z < 0 ) {  // for simplicity snow doesn't fall below ground
+        return;
+    }
+
+    level_cache &cache = get_cache( gridp.z );
+    level_cache *cache_above = gridp.z < OVERMAP_HEIGHT ? &get_cache( gridp.z + 1 ) : nullptr;
+    submap *sm = get_submap_at_grid( gridp );
+
+    if( !sm ) {
+        return;
+    }
+    if( sm->is_uniform && gridp.z != 0 ) {
+        // skip uniform submaps above the ground
+        return;
+    }
+
+    static constexpr float eps = 0.001f;
+
+    // convert once
+    const field_type_id fd_snow = ::fd_snow;
+    const point sm_origin = sm_to_ms_copy( gridp.xy() );
+
+    /**
+     * Converts snow level in mm to the field intensity level
+     * 1 < 10mm (0.39")
+     * 2 ≈ 10mm-160mm (6.3")
+     * 3 ≈ 160mm-810mm (31.9")
+     */
+    const auto level_to_intensity = [&]( float level_mm ) -> float {
+        return level_mm < eps ? 0 :
+        level_mm <= 10.f ? 1.f :
+        std::powf( level_mm / 10.f, 1.f / 4 );
```

Comment:
Seems like GNU libstdc++ does not properly handle `powf` and `ceilf`

```suggestion
        std::pow( level_mm / 10.f, 1.f / 4 );
```